File: packages/xyd-opensdk-ruby/src/project.ts
import type { OpensdkSpecJson } from '@xyd-js/opensdk-core';

import { snakeCase } from './naming';
import { rbString } from './rbwriter';

/**
 * Emit `<pkg>.gemspec`: a dependency-free gem manifest (stdlib-only SDK, so no
 * `add_dependency`). Marked user-owned scaffold by the emitter so regen never
 * clobbers a hand-edited gemspec. Gets NO ownership header (`.gemspec` is an
 * extension the orchestrator leaves untouched), so `frozen_string_literal`
 * stays a valid line-1 magic comment.
 */
export function renderGemspec(spec: OpensdkSpecJson, pkg: string): string {
  const info = spec.info;
  const summary = `Ruby client for the ${info.title} API`;
  const description = info.description?.trim() || summary;
  // Publish identity from spec.info (OpenAPI `info` or an sdk.json `publish`
  // override), falling back to the historical opensdk/MIT defaults.
  const author = info.contact?.name || 'opensdk';
  const license = info.license?.identifier || 'MIT';
  const lines = [
    '# frozen_string_literal: true',
    '',
    'Gem::Specification.new do |spec|',
    `  spec.name = ${rbString(pkg)}`,
    `  spec.version = ${rbString(info.version || '0.0.0')}`,
    `  spec.summary = ${rbString(summary)}`,
    `  spec.description = ${rbString(description)}`,
    `  spec.authors = [${rbString(author)}]`,
    `  spec.license = ${rbString(license)}`,
  ];
  if (info.homepage) lines.push(`  spec.homepage = ${rbString(info.homepage)}`);
  if (info.repository) lines.push(`  spec.metadata = { "source_code_uri" => ${rbString(info.repository)} }`);
  lines.push(
    '  spec.required_ruby_version = ">= 2.6.0"',
    '  spec.files = Dir["lib/**/*.rb", "*.gemspec"]',
    '  spec.require_paths = ["lib"]',
    'end',
  );
  return `${lines.join('\n')}\n`;
}

/**
 * Emit `lib/<pkg>.rb`: the gem entry point. Requires the stdlib transports the
 * runtime needs, then the generated files in dependency order (transport ->
 * models -> resources -> client). Fully generated (overwrite): the require list
 * tracks the emitted file set.
 */
export function renderEntrypoint(spec: OpensdkSpecJson, pkg: string): string {
  const requires = [
    'require "json"',
    'require "net/http"',
    'require "securerandom"',
    'require "time"',
    'require "uri"',
    '',
    `require_relative ${rbString(`${pkg}/transport`)}`,
    `require_relative ${rbString(`${pkg}/models`)}`,
  ];
  for (const r of spec.resources || []) {
    requires.push(`require_relative ${rbString(`${pkg}/resources/${snakeCase(r.name)}`)}`);
  }
  requires.push(`require_relative ${rbString(`${pkg}/client`)}`);
  return `${requires.join('\n')}\n`;
}

File: crates/xyd_opensdk_ruby/src/project.rs
//! Port of project.ts — the `.gemspec` + `lib/<pkg>.rb` entry point.

use serde_json::Value;

use crate::naming::snake_case;
use crate::writer::rb_string;

pub fn render_gemspec(spec: &Value, pkg: &str) -> String {
    render_gemspec_with(spec, pkg, "API")
}

/// CLI-mode gemspec: identical fields, but the summary names the CLI. The
/// file list (`lib/**/*.rb` + the gemspec itself) is already CLI-relevant-only
/// — no test extras or development dependencies are declared in either mode.
pub fn render_gemspec_cli(spec: &Value, pkg: &str) -> String {
    render_gemspec_with(spec, pkg, "CLI")
}

fn render_gemspec_with(spec: &Value, pkg: &str, what: &str) -> String {
    let info = spec.get("info");
    let title = info
        .and_then(|i| i.get("title"))
        .and_then(|t| t.as_str())
        .unwrap_or("");
    let summary = format!("Ruby client for the {title} {what}");
    let description = info
        .and_then(|i| i.get("description"))
        .and_then(|d| d.as_str())
        .map(|d| d.trim())
        .filter(|d| !d.is_empty())
        .unwrap_or(&summary)
        .to_string();
    let author = info
        .and_then(|i| i.get("contact"))
        .and_then(|c| c.get("name"))
        .and_then(|n| n.as_str())
        .filter(|n| !n.is_empty())
        .unwrap_or("opensdk");
    let license = info
        .and_then(|i| i.get("license"))
        .and_then(|l| l.get("identifier"))
        .and_then(|n| n.as_str())
        .filter(|n| !n.is_empty())
        .unwrap_or("MIT");
    let version = info
        .and_then(|i| i.get("version"))
        .and_then(|v| v.as_str())
        .filter(|v| !v.is_empty())
        .unwrap_or("0.0.0");

    let mut lines = vec![
        "# frozen_string_literal: true".to_string(),
        String::new(),
        "Gem::Specification.new do |spec|".to_string(),
        format!("  spec.name = {}", rb_string(pkg)),
        format!("  spec.version = {}", rb_string(version)),
        format!("  spec.summary = {}", rb_string(&summary)),
        format!("  spec.description = {}", rb_string(&description)),
        format!("  spec.authors = [{}]", rb_string(author)),
        format!("  spec.license = {}", rb_string(license)),
    ];
    if let Some(hp) = info
        .and_then(|i| i.get("homepage"))
        .and_then(|h| h.as_str())
        .filter(|h| !h.is_empty())
    {
        lines.push(format!("  spec.homepage = {}", rb_string(hp)));
    }
    if let Some(repo) = info
        .and_then(|i| i.get("repository"))
        .and_then(|r| r.as_str())
        .filter(|r| !r.is_empty())
    {
        lines.push(format!(
            "  spec.metadata = {{ \"source_code_uri\" => {} }}",
            rb_string(repo)
        ));
    }
    lines.push("  spec.required_ruby_version = \">= 2.6.0\"".to_string());
    lines.push("  spec.files = Dir[\"lib/**/*.rb\", \"*.gemspec\"]".to_string());
    lines.push("  spec.require_paths = [\"lib\"]".to_string());
    lines.push("end".to_string());
    format!("{}\n", lines.join("\n"))
}

pub fn render_entrypoint(spec: &Value, pkg: &str) -> String {
    let mut requires = vec![
        "require \"json\"".to_string(),
        "require \"net/http\"".to_string(),
        "require \"securerandom\"".to_string(),
        "require \"time\"".to_string(),
        "require \"uri\"".to_string(),
        String::new(),
        format!(
            "require_relative {}",
            rb_string(&format!("{pkg}/transport"))
        ),
        format!("require_relative {}", rb_string(&format!("{pkg}/models"))),
    ];
    if let Some(resources) = spec.get("resources").and_then(|r| r.as_array()) {
        for r in resources {
            let sn = snake_case(r.get("name").and_then(|n| n.as_str()).unwrap_or(""));
            requires.push(format!(
                "require_relative {}",
                rb_string(&format!("{pkg}/resources/{sn}"))
            ));
        }
    }
    requires.push(format!(
        "require_relative {}",
        rb_string(&format!("{pkg}/client"))
    ));
    format!("{}\n", requires.join("\n"))
}
